### PR TITLE
libroach: detected leaked iterators

### LINKED ISSUE #23138: *: add iterator leak test
Forgetting to close a RocksDB iterator causes all manner of subtle bugs: file descriptors run out, or the disk fills up, and often hours after the iterator was leaked. #23117 is the most recent example of this.

Since Go can't do us any favors here, we should teach `leaktest.AfterTest(t)` to check for leaked iterators so we can avoid this class of bug once and for all. We could also consider a flag to enable "time-bomb" iterators that explode if kept open for more than ~1m.

### PR DESCRIPTION
**Pull request by petermattis**
Keep track of the number of live iterators and assert when an engine is
closed that the value is 0.

Fixes #23138

Release note: None

### CODE REVIEW

**Comment by petermattis** [comment]
It appears that we have a dozen or so tests that leak iterators. Can't merge this PR until I track them all down and fix them.

File: c-deps/libroach/engine.cc [hunk 1]
```diff
@@ -23,6 +23,15 @@
 
 using namespace cockroach;
 
+DBEngine::~DBEngine() {
+  const int64_t n = iters.load();
+  if (n != 0) {
+    fprintf(stderr, "%lld leaked iterators\n", n);
```
Comment (benesch) [inline c-deps/libroach/engine.cc]: I think you need:

```
fprintf(stderr, "%" PRId64 " leaked iterators\n", n)
```

**Review by benesch — APPROVED** [review]
Code so far LGTM

**Comment by petermattis** [comment]
Ok, I figured out all but one of the failures: batches are closed before their iterations are destroyed. This is due to the fast path of closing batches when they commit and is difficult to change. Instead, I moved the code which performs the assertion to only run when the main rocksdb instance is closed, rather than on every close of a snapshot or batch.

`TestImportStmt` is still reporting 12 leaked iterators. I haven't determined if this is real or not. I'm wondering if I should move this iterator leak check from C++ to Go. Doing the check in Go would have the advantage that for race builds we could keep a stack trace to show where the leaked iterator was allocated. Much more difficult to do this from C++. 

Btw, I verified that this assertion caught the leak that Nikhil fixed recently, but it didn't help in tracking it down. Hence my interest in moving the check to Go.

**Comment by petermattis** [comment]
Added a second commit which fixes `TestImportStmt`: there was a real leak in the import `sstWriter` code, though the effects were much less serious than previous iterator leaks.

I also added disabled code which tracks the stacks of where iterators are allocated. This is expensive and thus disabled by default, but we can enable if the fast leak check ever fires again.

**Review by benesch — APPROVED** [review]

**Comment by benesch** [comment]
This is already paying dividends! How excellent.

**Comment by bdarnell** [comment]
:lgtm: 

---

Reviewed 14 of 14 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/engine/rocksdb.go, line 74 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23183#-L6PRK4d6gklXb77vvcB:-L6PRK4d6gklXb77vvcC:beylsfy) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e567a8a6ed9df3b115317b4abe47ca8a08ac82e/pkg/storage/engine/rocksdb.go#L74)):*
> ```Go
> // not indicate where the leak arose. The expensive checking tracks the stack
> // traces of every iterator allocated. DO NOT ENABLE in production code.
> const debugIteratorLeak = false
> ```

Should we turn this on automatically in race builds? Or is it enough that the assertions are always enabled and we can turn on the stack traces manually?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23183#-:-L6PRX59fbhffVByzX6b:bg7igkv)*

**Comment by petermattis** [comment]
Review status: 13 of 15 files reviewed at latest revision, 2 unresolved discussions.

---

*[c-deps/libroach/engine.cc, line 29 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23183#-L6O2G3Mi7St_tBNxLol-r1-29:-L6Riict7jskMr4g7V9f:bs7wumb) ([raw file](https://github.com/cockroachdb/cockroach/blob/58f1faa27bdf589279155976d7085f918dab55e4/c-deps/libroach/engine.cc#L29)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

I think you need:

```
fprintf(stderr, "%" PRId64 " leaked iterators\n", n)
```
</blockquote></details>

Done.

---

*[pkg/storage/engine/rocksdb.go, line 74 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23183#-L6PRK4d6gklXb77vvcB:-L6Rix8TCZfBLSM5RlcN:b-wj35wv) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e567a8a6ed9df3b115317b4abe47ca8a08ac82e/pkg/storage/engine/rocksdb.go#L74)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Should we turn this on automatically in race builds? Or is it enough that the assertions are always enabled and we can turn on the stack traces manually?
</blockquote></details>

It felt sufficient to enable it manually. I'm anxious about making our race builds even slower.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23183)*

**Comment by petermattis** [comment]
Review status: 14 of 15 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[c-deps/libroach/engine.cc, line 29 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23183#-L6O2G3Mi7St_tBNxLol-r1-29:-L6S-K55xjFq3He_tNFd:bqlwnk5) ([raw file](https://github.com/cockroachdb/cockroach/blob/58f1faa27bdf589279155976d7085f918dab55e4/c-deps/libroach/engine.cc#L29)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Using `PRId64` failed when building under the builder. Reverted to `%lld` which we use in at least one other place in `libroach`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23183)*

**Comment by benesch** [comment]
How odd. (Your first build using %lld also failed FWIW.)

**Comment by petermattis** [comment]
> How odd. (Your first build using %lld also failed FWIW.)

The most recent build succeeded and we're definitely using `%lld` elsewhere. What to do?

**Comment by petermattis** [comment]
Review status: 12 of 15 files reviewed at latest revision, 2 unresolved discussions.

---

*[c-deps/libroach/engine.cc, line 29 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23183#-L6O2G3Mi7St_tBNxLol-r1-29:-L6U71jfdvRSLxLdDgYq:btllbnp) ([raw file](https://github.com/cockroachdb/cockroach/blob/58f1faa27bdf589279155976d7085f918dab55e4/c-deps/libroach/engine.cc#L29)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Using `PRId64` failed when building under the builder. Reverted to `%lld` which we use in at least one other place in `libroach`. 
</blockquote></details>

Back to using `PRId64` (again). How many times can I change this line in one PR?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23183)*

**Pull request merged**